### ISSUE TITLE
Reindex from Remote to support auth without username and password

### ISSUE THREAD

**Issue by ankikala** [elastic/elasticsearch#58396]

**Comment by tvernum** [comment]
> this currently supports username and password as only mechanism to authenticate

You can also authenticate via PKI (TLS client certificates) by configuring [`reindex.ssl.key` or `reindex.ssl.keystore.path`](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#reindex-ssl).

> using auth like saml, kerberos, oidc (and IAM)

The issue is that there is no standard way to provide SAML or OIDC credentials over a rest connection. Both SAML and OIDC are primarily used as browser based SSO protocols, and they require an interactive user interface in order to facilitate authentication. There isn't a easy or standard way to make that work between two backend services like ES.

Kerberos is theoretically possible (for example, we support Kerberos authentication from Hadoop) but it wouldn't work the way you would expect - it would require that the ES instance was running with a Kerberos identity. It wouldn't use the end-user's identity.

> If required, I can probably help out in building the support

If you have specific thoughts about _how_ that would work, I'd be interested in hearing them, because I don't think that what you are asking for is feasible.

**Comment by ankikala** [comment]
I agree that there doesn't seem to be a generalized solution for this. In particular, not sure if there can even be any solution for SSO based authentication. 
On a high level, I was thinking about simply adding plugin support to intercept outgoing REST calls and adding the relevant headers. What exactly needs to be done for each auth type can be handled by different plugins. I guess I was thinking about this more from IAM perspective but I agree, wouldn't work with SAML and OIDC(and barely for kerberos). 

Just curious, what is the current recommendation for doing remote reindex on such domains?

**Comment by tvernum** [comment]
> Just curious, what is the current recommendation for doing remote reindex on such domains?

All of Elastic's deployments/distributions of ES (whether cloud hosted, or installed on premises) support some sort of username+password authentication. We don't really run into cases where we people need reindex to work across a different authentication scheme.

**Comment by PhaedrusTheGreek** [comment]
Another consideration here would be the use of an API Key via HTTP Header

e.g.:  

```
curl -H "Authorization: ApiKey VnVhQ2ZHY0JDZGJrUW0tZTVhT3g6dWkybHAyYXhUTm1zeWFrdzl0dk5udw==" http://localhost:9200/_cluster/health

```
